src/Queue/DoctrineQueue.php

```diff
@@ -182,19 +225,162 @@ public function pop(array $options = []): ?JobInterface
+
+                ->set('job.workerId = :workerId')
+                ->setParameter('workerId', $this->workerId)
+
+                ->where('job.status = :oldStatus')
+                ->setParameter('oldStatus', static::STATUS_PENDING)
+
+                ->andWhere('job.queue = :queue')
+                ->setParameter('queue', $this->getName())
+
+                ->andWhere('scheduled <= :scheduled')
+                ->setParameter('scheduled', $this->now->format('Y-m-d H:i:s.u'))
+
+                ->addOrderBy('priority', 'ASC')
+                ->addOrderBy('scheduled', 'ASC')
+
+                ->setMaxResults(1);
+
+            $result = $this->connection->executeQuery(
+                $queryBuilder->getSQL(),
+                $queryBuilder->getParameters(),
+                $queryBuilder->getParameterTypes()
+            );
+
+            $rowCount = $result->rowCount();
+
+            if ($rowCount === 0) {
+                // No row was updated -> no job ready for us!
+                return null;
+            }
+
+            if ($rowCount > 1) {
+                // That should not be possible, but what if ...? Reset them...
+                $queryBuilder = $this->connection->createQueryBuilder();
+                $queryBuilder
+                    ->update($this->options->getTableName(), 'job')
+
+                    ->set('job.status = :newStatus')
+                    ->setParameter('newStatus', static::STATUS_PENDING)
+
+                    ->set('job.workerId = NULL')
+
+                    ->where('job.status = :oldStatus')
+                    ->setParameter('oldStatus', static::STATUS_ALLOCATING)
+
+                    ->andWhere('job.workerId = :workerId')
+                    ->setParameter('workerId', $this->workerId)
+
+                    ->andWhere('job.queue = :queue')
+                    ->setParameter('queue', $this->getName());
+
+                $this->connection->executeQuery(
+                    $queryBuilder->getSQL(),
+                    $queryBuilder->getParameters(),
+                    $queryBuilder->getParameterTypes()
+                );
+
+                throw new LogicException(
+                    'More than 1 job was assigned to our worker despite using UPDATE with LIMIT 1!'
+                );
+            }
+
+            $queryBuilder = $this->connection->createQueryBuilder();
+            $queryBuilder
+                ->select('*')
+                ->from($this->options->getTableName())
+
+                ->where('status = :oldStatus')
+                ->setParameter('oldStatus', static::STATUS_ALLOCATING)
+
+                ->andWhere('workerId = :workerId')
+                ->setParameter('workerId', $this->workerId);
+
+            $result = $this->connection->executeQuery(
+                $queryBuilder->getSQL(),
+                $queryBuilder->getParameters(),
+                $queryBuilder->getParameterTypes()
+            );
+
+            if ($row = $result->fetch()) {
```

Comment:
I think the flow is more readable if you use an early return here:

```php
$row = $result->fetch();
if (!$row) {
    retrun null;
}

$queryBuilder = 
```

And that a `return null`  at the bottom of this method.